### DISCUSSION TITLE
Neighbor Search

### DISCUSSION THREAD

**Discussion by lobnaOmar20** [project-gemmi/gemmi#279] [Q&A] [answered]
using Neighbour Search, I want to get bond lengths or interatomic distances: 

```
small = gemmi.read_small_structure('supercell_i00002_w8.cif')
si_site = small.sites[0]
ns = gemmi.NeighborSearch(small,4.0).populate()
ns
```

<gemmi.NeighborSearch with grid 3, 4, 3>

```
for mark in ns.find_site_neighbors(si_site, min_dist=0.1):
    site = mark.to_site(small)
    print(site.label, 'symmetry op #%d' %mark.image_idx)
```

gives me:

Si3 symmetry op #0
O3 symmetry op #0
O9 symmetry op #0
O8 symmetry op #0
O6 symmetry op #0
Si5 symmetry op #0
O4 symmetry op #0
O8 symmetry op #0
Si3 symmetry op #0
O3 symmetry op #0
O9 symmetry op #0
Al2 symmetry op #0
O10 symmetry op #0
O2 symmetry op #0
O7 symmetry op #0

which is partially right and partially wrong 

The distances from Si3 to the nearest neighbor is between 1.60 A and 1.70 A. 
and therefore Si3 has O9 O7 O3 O2 neighbors
whereas the solution with gemmi suggests  O9 O6 O3 O8

I don't understand what is wrong? 

I have P1 structure and fractional coordinates

**Comment by wojdyr** [comment]
From this description I don't know if it's right or wrong.
If you could minimize this example to the smallest set of atoms that demonstrate the problem, and include a CIF file with these atoms, I'll have a look. For example, leave only two atoms that should be neighbors but are not shown by NeighborSearch.

**Reply by lobnaOmar20** [reply]
Thank you so much for your quick reply. 

So I have my cif file attached as txt because the website can't take cif files. I want to get the bond length between Si1 and O6, Si1 and O3, Si1 and O4, Si1 and O9. The tetrahedral site S1 is surrounded by O6 O3 O4 O9 therefore I want to get the interatomic distances or the bond length between both. 


[supercell_i00002_w8.txt](https://github.com/project-gemmi/gemmi/files/12405960/supercell_i00002_w8.txt)

**Reply by lobnaOmar20** [reply]
when I try to use gemmi Neighbor search I get this, even though I search first for Si1 but I don't get Si1. In addition, Si3 is surrounded by O9,O7,O3,O2 but in the solution it shows O3 O9 O8 O6

Si3 symmetry op #0
O3 symmetry op #0
O9 symmetry op #0
O8 symmetry op #0
O6 symmetry op #0
Si5 symmetry op #0
O4 symmetry op #0
O8 symmetry op #0
Si3 symmetry op #0
O3 symmetry op #0
O9 symmetry op #0
Al2 symmetry op #0
O10 symmetry op #0
O2 symmetry op #0
O7 symmetry op #0

**Comment by wojdyr** [comment, marked as answer]
> The distances from Si3 to the nearest neighbor is between 1.60 A and 1.70 A.

and you search in the radius of 4.0A, getting also the second nearest neighbors.
Add `max_dist=2.0` to ns.find_site_neighbors().

If I limit the search to 2.0 I get:

find neighbors of Si1 <gemmi.Fractional(0.6784, 0.6701, 0.0983)>
O3 symmetry op #0 <gemmi.Fractional(0.685, 0.6573, 0.8891)> 1.619205318374039
O6 symmetry op #0 <gemmi.Fractional(0.5806, 0.6616, 0.0572)> 1.6229646008471021
O4 symmetry op #0 <gemmi.Fractional(0.7353, 0.6063, 0.2558)> 1.6235125698772466
O9 symmetry op #0 <gemmi.Fractional(0.7094, 0.7544, 0.1882)> 1.639821180192079

Note that the distance to O3 is calculated through the PBC. It can be calculated in gemmi with:

    d = small.cell.orthogonalize((a.fract - b.fract).wrap_to_zero()).length()

**Reply by lobnaOmar20** [reply]
Thank you very much :)

**Reply by lobnaOmar20** [reply]
how to cite gemmi library or you ??

**Reply by lobnaOmar20** [reply]
okay found it in one of the older questions

**Reply by lobnaOmar20** [reply]
what is op #0 in <O3 symmetry op #0 > ?? 
and how did you output the gemmi.Fractional() after each oxygen atom like that ??

I am trying to input to gemmi.Fractional() the positions of atoms but it is not accepting them from the arrays. I will have to add them explicitly ??

**Reply by lobnaOmar20** [reply]
when I try also to add the positions explicitly, and use the d = small.cell.orthogonalize((a.fract - b.fract).wrap_to_zero()).length()

S1 = gemmi.Fractional(0.6784, 0.6701, 0.0983)
O3 = gemmi.Fractional(0.685, 0.6573, 0.8891)

d = small.cell.orthogonalize((S1 - O3).wrap_to_zero()).length()

I get 
gemmi.Fractional' object has no attribute 'wrap_to_zero'

**Reply by wojdyr** [reply]
wrap_to_zero was added before v0.6.0.

here is the full script that I used for testing:
```py
import sys
import gemmi

small = gemmi.read_small_structure(sys.argv[1])
ns = gemmi.NeighborSearch(small, 4.0).populate()
print(ns)
x = small.sites[0]
print('find neighbors of', x.label, x.fract)
for mark in ns.find_site_neighbors(x, min_dist=0.1, max_dist=2.0):
  site = mark.to_site(small)
  d = small.cell.orthogonalize((x.fract - site.fract).wrap_to_zero()).length()
  print(site.label, 'symmetry op #%d' % mark.image_idx, site.fract, d)
```

**Reply by lobnaOmar20** [reply]
that is awesome thank you soooooo much.

**Reply by lobnaOmar20** [reply]
AttributeError: 'gemmi.Fractional' object has no attribute 'wrap_to_zero'
again ^_^ I think as you said due to the version. should I install the older version ?

**Reply by wojdyr** [reply]
What version do you use?

**Reply by lobnaOmar20** [reply]
I use (0.5.7) ohh then may be I do need an update... 
I removed wrap to zero and it worked but with only one wrong distance between Si1 and O3. Take a look: 

<gemmi.NeighborSearch with grid 3, 4, 3>
find neighbors of Si1 <gemmi.Fractional(0.6784, 0.6701, 0.0983)>
O3 symmetry op #0 <gemmi.Fractional(0.685, 0.6573, 0.8891)> 5.806881296696083
O6 symmetry op #0 <gemmi.Fractional(0.5806, 0.6616, 0.0572)> 1.6229646008471021
O4 symmetry op #0 <gemmi.Fractional(0.7353, 0.6063, 0.2558)> 1.6235125698772466
O9 symmetry op #0 <gemmi.Fractional(0.7094, 0.7544, 0.1882)> 1.639821180192079

**Reply by lobnaOmar20** [reply]
I upgraded to the latest version and It worked. Thank you sooooo much.

**Reply by lobnaOmar20** [reply]
You don't know how much Gemmi library helped me with my work. thank youuuuu
